StudentPlan.py

```diff
@@ -23,50 +32,97 @@ def accommodate(self, course, correq=False):
         return min_position + 1 if not correq else course.position
 
-    def add_semester(self, position):
-        semester = Semester(max_credits=self.max_credits, position=position)
+    def add_semester(self, position, past=False):
+        semester = Semester(max_credits=self.max_credits, position=position, past=past)
         self.semesters.append(semester)
         return semester
 
-    def bops(self, course, min_position):
-        if min_position >= len(self.semesters):
+    def force_accommodate(self, position, course):
+        if not self.get_semester(position):
+            new_semester = self.add_semester(position)
+            new_semester.add_course(course)
+
+    def get_semester(self, position):
+        for semester in self.semesters:
+            if semester.position == position:
+                return semester
+
+    def set_to_earliest_possible_semester(self, course, min_position):
+        if not self.get_semester(min_position):
             self.add_semester(position=min_position)
+        self.semesters = sorted(self.semesters, key=lambda s: s.position,)
         for semester_index in range(min_position, len(self.semesters)):
-            if not self.semesters[semester_index].is_full:
-                if self.semesters[semester_index].credit_hours + course.credit_hours <= self.semesters[semester_index].max_credits:
-                    self.semesters[semester_index].add_course(course)
-                    return semester_index
+            if self.semesters[semester_index].course_valid(course):
+                self.semesters[semester_index].add_course(course)
+                return semester_index
         semester = self.add_semester(position=len(self.semesters))
         semester.add_course(course)
         return semester.position
 
 
 class Semester:
-
     is_full = False
     credit_hours = 0
 
-    def __init__(self, max_credits, position):
+    def __init__(self, max_credits, position, past=False):
         self.max_credits = max_credits
         self.position = position
         self.courses = list()
+        self.past = past
+
+    def course_valid(self, course):
+        if not self.is_full and not self.past and self.credit_hours + course.get_credit_hours() <= self.max_credits:
+            if course.season == 2:
+                return self.position % 2 == 0
+            elif course.season == 1:
+                return self.position % 2 == 1
+            else:
+                return True
 
     def add_course(self, course):
         course.position = self.position
         self.courses.append(course)
-        self.credit_hours += course.credit_hours
+        if course.lab:
+            course.lab.position = self.position
+            self.courses.append(course.lab)
+        self.credit_hours += course.get_credit_hours()
         self.is_full = self.credit_hours >= self.max_credits
 
+    def get_year(self):
+        return int(self.position/2) if self.position % 2 == 0 else self.position//2 + 1
```

Comment:
int(self.position/2) --> self.position//2

Reply:
this is to eliminate the .0 at the end of the number.